pkg/secrets/watch.go

```diff
@@ -0,0 +1,266 @@
+	"k8s.io/client-go/kubernetes"
+)
+
+// WatchSPConfig configures access to the Kubernetes API server.
+// TODO(TheSpiritXIII): https://github.com/GoogleCloudPlatform/prometheus-engine/issues/867
+type WatchSPConfig struct {
+	ClientConfig
+}
+
+func (c *WatchSPConfig) newProvider(ctx context.Context, opts ProviderOptions) (*watchProvider, error) {
+	client, err := c.ClientConfig.client()
+	if err != nil {
+		return nil, err
+	}
+	return newWatchProvider(ctx, opts.Logger, client)
+}
+
+type secretWatcher struct {
+	// Add, Update and Remove are synchronous. We need to lock everything but `refCount`.
+	mu       sync.Mutex
+	w        watch.Interface
+	s        *corev1.Secret
+	refCount uint
+	done     bool
+}
+
+func newWatcher(ctx context.Context, logger log.Logger, client kubernetes.Interface, config *KubernetesSecretConfig) (*secretWatcher, error) {
+	watcher := &secretWatcher{
+		refCount: 1,
+		done:     false,
+	}
+
+	if err := watcher.start(ctx, client, config); err != nil {
+		return nil, err
+	}
+
+	go func() {
+		for {
+			select {
+			case e, ok := <-watcher.w.ResultChan():
+				if ok {
+					watcher.update(logger, e)
+					continue
+				}
+
+				// If the application shutdown, we don't care about cleanup.
+				if ctx.Err() != nil {
+					watcher.mu.Lock()
+					defer watcher.mu.Lock()
+					watcher.s = nil
+					return
+				}
+				// If closed unintentionally (i.e. network issues), try and restart it.
+				for {
+					ok, err := watcher.restart(ctx, client, config)
+					if ok {
+						return
+					}
+					// If an error occurred trying to watch, keep retrying.
+					if err == nil {
+						break
+					}
+					_ = logger.Log("msg", "unable to restart secret watcher", "err", err, "namespace", watcher.s.Namespace, "name", watcher.s.Name)
+				}
+			case <-ctx.Done():
+				// The application shutdown, we don't care about cleaning up.
+				watcher.close()
+				return
+			}
+		}
+	}()
+
+	return watcher, nil
+}
+
+func (w *secretWatcher) update(logger log.Logger, e watch.Event) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	switch e.Type {
+	case watch.Modified, watch.Added:
+		secret := e.Object.(*corev1.Secret)
+		w.s = secret
+	case watch.Deleted:
+		w.s = nil
+	case watch.Bookmark:
+		// Disabled explicitly when creating the watch interface.
+	case watch.Error:
+		//nolint:errcheck
+		logger.Log("msg", "watch error event", "namespace", w.s.Namespace, "name", w.s.Name)
+	}
+}
+
+func (w *secretWatcher) secret(config *KubernetesSecretConfig) Secret {
+	fn := SecretFn(func(_ context.Context) (string, error) {
+		w.mu.Lock()
+		defer w.mu.Unlock()
+		if w.s == nil {
+			return "", fmt.Errorf("secret %s/%s not found", config.Namespace, config.Name)
+		}
+		return getKey(w.s, config.Key)
+	})
+	return &fn
+}
+
+func (w *secretWatcher) start(ctx context.Context, client kubernetes.Interface, config *KubernetesSecretConfig) error {
+	var err error
+	w.w, err = client.CoreV1().Secrets(config.Namespace).Watch(ctx, metav1.ListOptions{
```

Comment:
question (non-blocking): I'm curious, did you see if it was feasible or beneficial to share a watch among the secrets, instead of initializing a watch per secret?

Reply:
No, because we would need RBAC permissions to watch all secrets in the entire cluster.